runtime/cudaq/qis/qubit_qis.h

```diff
@@ -147,31 +136,158 @@ void oneQubitApplyControlledRange(QubitRange &ctrls, qubit &target) {
+
+  // Get the number of input qubits
+  constexpr std::size_t nArgs = sizeof...(QubitArgs);
+
+  // Map the qubits to their unique ids and pack them.
+  std::vector<QuditInfo> quditInfos{qubitToQuditInfo(args)...};
+  std::vector<bool> qubitIsNegated{qubitIsNegative(args)...};
+
+  // If we are here, then `mod` must be control or adjoint. Extract the controls
+  // and the target
+  std::vector<QuditInfo> controls(quditInfos.begin(),
+                                  quditInfos.begin() + nArgs - 1);
+
+  // If we have controls, check if any of them are negative controls, and if so
+  // apply an x.
+  if (!controls.empty())
+    for (std::size_t i = 0; i < controls.size(); i++)
+      if (qubitIsNegated[i])
+        getExecutionManager()->apply("x", {}, {}, {controls[i]});
+
+  // Apply the gate
+  getExecutionManager()->apply(gateName, {}, controls, {quditInfos.back()},
+                               /*adjoint=*/false);
+
+  // If we did apply any X ops for a negative control, we need to reverse it.
+  if (!controls.empty())
+    for (std::size_t i = 0; i < controls.size(); i++)
+      if (qubitIsNegated[i]) {
+        getExecutionManager()->apply("x", {}, {}, {controls[i]});
+        // fold expression which will reverse the negation
+        (
+            [&] {
+              if (args.is_negative())
+                args.negate();
+            }(),
+            ...);
+      }
+}
+
+template <typename QuantumOp, typename QubitRange>
+void oneQubitApplyControlledRange(QubitRange &ctrls, qubit &target) {
+  // Get the name of the operation
+  auto gateName = QuantumOp::name();
+
+  // Map the input control register to a vector of QuditInfo
+  std::vector<QuditInfo> controls;
+  std::transform(ctrls.begin(), ctrls.end(), std::back_inserter(controls),
+                 [](auto &q) { return cudaq::qubitToQuditInfo(q); });
+
+  // Apply the gate
+  getExecutionManager()->apply(gateName, {}, controls,
+                               {cudaq::qubitToQuditInfo(target)});
+}
+
+#define CUDAQ_QIS_ONE_TARGET_QUBIT_(NAME)                                      \
+  namespace types {                                                            \
+  struct NAME {                                                                \
+    inline static const std::string name{#NAME};                               \
+  };                                                                           \
+  }                                                                            \
+  template <typename... QubitArgs>                                             \
+  void NAME(QubitArgs &...args) {                                              \
+    oneQubitApply<qubit_op::NAME##Op>(args...);                                \
+  }                                                                            \
+  template <typename... QubitArgs>                                             \
+  void c##NAME(QubitArgs &...args) {                                           \
```

Comment:
Do we need to provide an equivalence for adjoint, e.g., `sdg` and `tdg`?

Reply:
This allows `cOP(anycontrols..., target)` which I don't think is what we want.

Reply:
See change c16e5ffa9